### PR TITLE
Port ClinSeq to use WorkflowBuilder

### PR DESCRIPTION
**Pull request by susannasiebert**
This PR makes ClinSeq use WorkflowBuilder. It also includes some refactoring of the _resolve_workflow_for_build subroutine by split the various parts of the workflow into separate subroutines for clarity (hopefully).

I ran a ClinSeq test build in a custom checkout and the workflows look identical. @jasonwalker80, you mentioned that you were working on a ClinSeq model that would include all possible steps. It might be worthwhile to run a build of that model on this PR to make sure that all steps are still executed.

@acoffman, @gatoravi , @jasonwalker80 Feel free to push additional commits to this PR with any improvements that might be necessary while I'm on vacation.

### CODE REVIEW

**Comment by susannasiebert** [comment]
I apologize to the people that I accidentally tagged incorrectly.

File: lib/perl/Genome/Model/Build/ClinSeq/InputBuilds.pm [hunk 1]
```diff
@@ -30,4 +30,9 @@ sub resolve_rnaseq_builds {
   $rnaseq_builds->{$normal_build->id}{type} = 'rnaseq' if $normal_build;
 }
 
+sub wgs_exome_build {
```
Comment (tmooney) [inline lib/perl/Genome/Model/Build/ClinSeq/InputBuilds.pm]: This name is a little strange.  The other `..._build`s return builds, but this returns a boolean.  I see how it's programmatically convenient in the rest of this commit, though.

File: lib/perl/Genome/Model/ClinSeq.pm [hunk 2]
```diff
@@ -2027,6 +1969,71 @@ sub make_circos_plot_op {
     return $make_circos_plot_op;
 }
 
+sub converge_snv_indel_report_op {
+    my $self = shift;
+    my $workflow = shift;
+    my $iterator = shift;
```
Comment (tmooney) [inline lib/perl/Genome/Model/ClinSeq.pm]: This variable seems to hold an index rather than an iterator.

File: lib/perl/Genome/Model/ClinSeq.pm [hunk 3]
```diff
@@ -1062,85 +1062,27 @@ sub _resolve_workflow_for_build {
     #Converge SnvIndelReport
     my @converge_snv_indel_report_ops;
     if ($build->wgs_build || $build->exome_build) {
+        my %variant_sources_ops = (
+            exome => $exome_variant_sources_op,
+            wgs   => $wgs_variant_sources_op,
+        );
         #Create a report for each $bq $mq combo.
         my $iterator = List::MoreUtils::each_arrayref([1 .. @$mqs], $mqs, $bqs);
         while (my ($i, $mq, $bq) = $iterator->()) {
```
Comment (tmooney) [inline lib/perl/Genome/Model/ClinSeq.pm]: It looks like `$mq` and `$bq` are unused inside this loop.

File: lib/perl/Genome/Model/ClinSeq.pm [hunk 4]
```diff
@@ -2034,6 +1981,56 @@ sub converge_snv_indel_report_op {
     return $converge_snv_indel_report_op;
 }
 
+sub create_mutation_spectrum_op {
+    my $self = shift;
+    my $workflow = shift;
+    my $sequencing_type = shift;
+    my $iterator = shift;
```
Comment (tmooney) [inline lib/perl/Genome/Model/ClinSeq.pm]: Same comment about `$iterator` being an index.

File: lib/perl/Genome/Model/ClinSeq.pm [hunk 5]
```diff
@@ -645,33 +645,18 @@ sub _resolve_workflow_for_build {
...
-    }
-    #TophatJunctionsAbsolute - Run tophat junctions absolute analysis on tumor
-    my $tumor_tophat_junctions_absolute_op;
-    if ($build->tumor_rnaseq_build) {
-        $tumor_tophat_junctions_absolute_op = $self->tophat_junctions_absolute_op($workflow, 'tumor');
-    }
-
     #CufflinksExpressionAbsolute - Run cufflinks expression absolute analysis on normal
     if ($build->normal_rnaseq_build) {
+        my $normal_tophat_junctions_absolute_op = $self->tophat_junctions_absolute_op($workflow, 'normal');
```
Comment (tmooney) [inline lib/perl/Genome/Model/ClinSeq.pm]: This variable (and the one right below it) are unused.

Comment (jasonwalker80) [inline lib/perl/Genome/Model/ClinSeq.pm]: I presume this is why the scope of this variable is limited to the conditional block.  If the normal is defined then the downstream annotation step is linked to the differential_expression rather than these normal-only steps.  Should these operations be used elsewhere in the workflow?

I'm running a test build now with the normal RNA-seq defined.  The test @susannasiebert was using did not have a normal rnaseq input and may not have covered this case.

File: lib/perl/Genome/Model/Build/ClinSeq/InputBuilds.pm (hunk 1 above)
Comment (jasonwalker80) [inline lib/perl/Genome/Model/Build/ClinSeq/InputBuilds.pm]: Are there any suggestions for improving the consistency between these subroutines?  I'm not entirely sure what the rest of ClinSeq expects these to return.  There is a future task to refactor how ClinSeq resolves all of these input properties:
https://jira.gsc.wustl.edu/browse/CI-109

File: lib/perl/Genome/Model/ClinSeq.pm (hunk 5 above)
Comment (jasonwalker80) [inline lib/perl/Genome/Model/ClinSeq.pm]: These cufflinks steps ran successfully in a test ClinSeq model.  Are you suggesting that both variables should be removed since they are not used elsewhere?

>  7118929   execute  succeeded  2016-03-31T11:03:49       00:02:14           . Performing cufflinks expression absolute analysis for normal sample
>   7118933   execute  succeeded  2016-03-31T11:03:51       00:02:23           . Performing cufflinks expression absolute analysis for tumor sample

Comment (tmooney) [inline lib/perl/Genome/Model/ClinSeq.pm]: Yeah, I'm only suggesting that, since the return values are not used for anything, assigning them to `my` variables isn't necessary.

File: lib/perl/Genome/Model/Build/ClinSeq/InputBuilds.pm (hunk 1 above)
Comment (susannasiebert) [inline lib/perl/Genome/Model/Build/ClinSeq/InputBuilds.pm]: I agree with @jasonwalker80 that this is out of scope for this PR. I do think we should make the subroutine names consistent once we do the bigger ClinSeq refactor.

**Comment by susannasiebert** [comment]
@tmooney I pushed a few commits to address your suggestions.

File: lib/perl/Genome/Model/ClinSeq.pm [hunk 6]
```diff
@@ -944,8 +944,8 @@ sub _resolve_workflow_for_build {
             wgs   => $wgs_variant_sources_op,
         );
         #Create a report for each $bq $mq combo.
-        my $iterator = List::MoreUtils::each_arrayref([1 .. @$mqs], $mqs, $bqs);
-        while (my ($i, $mq, $bq) = $iterator->()) {
+        my $iterator = List::MoreUtils::each_arrayref([1 .. @$mqs]);
+        while (my ($i) = $iterator->()) {
```
Comment (tmooney) [inline lib/perl/Genome/Model/ClinSeq.pm]: These two lines could be simplified to: `for (1..@$mqs) {`.

**Comment by tmooney** [comment]
+1 :hamster:

File: lib/perl/Genome/Model/ClinSeq.pm (hunk 6 above)
Comment (susannasiebert) [inline lib/perl/Genome/Model/ClinSeq.pm]: Duh, of course. Fixed!

**Comment by jasonwalker80** [comment]
+1
